File: as.rb
#!/usr/bin/env ruby -w

__version__ = "emu8086 assembler 0.0.0(alpha)"


require './asclass'
class Assembler

  # initialize the Assembler instance
  def initialize(file,error = nil)

    @filename = "-"
    if file.class == String
      @content = file
    elsif file.class <= IO
      @content = file.read
      @filename = file.path if file.class == File
    end

    # File content by line
    @content = @content.split("\n")

    # Rows of file
    @rows = @content.length
    @line = 0

    # Exception
    @error = error || []
    @errorflag = false # flag true when disable to generate the code

  end

  prepend AssemblerClasses

  # procedure of assembling
  def assemble
    # initialize
    @code = {
      "main" => {
        :codes => [],
        :name  => "main",
        :type  => :code,
        :label => {}
      }
    }
    @segment = "main"
    @offset = 0
    label = nil

    # 1st: parse the code
    @rows.times do |i|
      @line = i

      # except comment `#` and `;`
      @content[@line] =~ /^((?:\'(?:(?:\\.)|[^\.])*?\'|[^;#])*)(?:[;#](.*))*$/
      # $1 = code without comment, $2 = comment
      code_string = $1
      comment = $2

      case code_string
      when /^(?:\s*(\w*):?\s+)*(segment|ends|db|dw)\s+(.*?)$/i # segment or data
        # $1 = label, $2 = pseudo instruction, $3 = options
        
        case $2.downcase
        when 'segment'
        when 'ends'
        else
          l = $1.strip.empty? ? nil : $1
          warning "label #{label} at offset #{@offset} has been redefine, the previous has been ignored" if label && l
          label = l if l
          code = $2
          para = $3
        end
      when /^(?:\s*(\w+):)*\s*(?:(\w+)(\s+.*?)?)?$/ # parse label and code
        # $1 = label, $2 = code, $3 = options
        warning "label #{label} at offset #{@offset} has been redefine, the previous has been ignored" if label && $1
        label = $1 if $1
        code = $2
        para = $3
      when /^\s*(%\w+)(\s+.*?)?/ # marco
        code = $1
        para = $2
      else
        fatal "parse error"
      end

      if label
        # label case sensitive
        unless @code[@segment][:label][label]
          @code[@segment][:label][label] = @offset
        else
          error "label #{label} redefine" unless @code[@segment][:label][label] == @offset
        end
      end
      
      if code
        code = SingletonCode.new(code, para, label, comment)
        code.apply(@@code_set, self)
        # offset += code.bytes
        # code.getready
        label = nil
      end

    end

    # 2nd: expression and marco

    # @code.each_pair do |key,value|
    #   value[:code].each do |code|
    #     # code.getready!
    #   end
    # end
    
  end
  
  # Code Exporting

  def binary

  end

  def code

  end
  
  # exception deal
  ["error", "warning", "fatal"].each do |p|
    define_method p.to_sym do |str|
      @error << "#{p}[#{@line + 1}]: #{str}\n"
      case p
      when "fatal" then throw "fatal error and halt"
      when "error" then @errorflag = true
      end
    end
  end

end

# main console program
if __FILE__ == $0
  STDERR.puts "Ruby 80x86 assembler"
  require 'optparse'
  options = {}
  cmd = OptionParser.new do |opts|
    opts.banner = "Usage: #{$0} [options]"

    opts.on("-v", "--version", "Show the version") do |v|
      STDERR.puts __version__
      exit
    end

    opts.on_tail("-h", "--help", "Show help") do
      STDERR.puts opts
      exit
    end

  end

  begin
    cmd.parse!
  rescue => exception
    STDERR.puts exception
    STDERR.puts cmd
    exit
  end
  
  if ARGV[0]
    as = Assembler.new(ARGV[0] == "-" ? STDIN : File.open(ARGV[0]), STDERR)
    as.assemble
    require 'pry'
    pry
    # print as.binary
  else
    STDERR.puts "no input file, halt"
    exit
  end
  
end

File: lib/as.js
const {Element, SingletonCode} = require('./asclass.js')
class Assembler {
  constructor(code,opts) {
    this.options = opts || {}
    this.content = code.split("\n")
    this.rows = content.length
    this.errorlist = []
    this.line = 0
  }

  assemble() {
    this.code = {
      "main": {
        codes: [],
        name: "code",
        label: {}
      },
      "data": {
        codes: [],
        name: "data",
        label: {}
      },
      "stack": {
        codes: [],
        name: "stack",
        label: {}
      },
      "extra": {
        codes: [],
        name: "extra",
        label: {}
      }
    }
    this.datalabel = {}
    this.segment = "main"
    this.offset = 0
    let label, code, para

    // parse the code

    ;[...Array(this.rows).keys()].forEach((i) => {
      this.line = i

      let match = this.content[this.line].match(/^((?:\'(?:(?:\\.)|[^\.])*?\'|[^;#])*)(?:[;#](.*))*$/)
      let code_string = match[1]
      let comment = match[2]

      if (match = code_string.match(/^(?:\s*(\w*):?\s+)*(segment|ends|db|dw)\s+(.*?)$/i)) {
        let pseudo = match[2]
        switch (pseudo) {
          case 'segment':
            break;
          case 'ends':
            break;
          default:
            let l = (match[1].trim() === '') ? void(0) : match[1];
            (label && l) && this.warning(`label ${label} at offset ${this.offset} has been redefine, the previous has been ignored`);
            l && (label = l);
            code = match[2];
            para = match[3];
            break;
        }
      }
      else if (match = code_string.match(/^(?:\s*(\w+):)*\s*(?:(\w+)(\s+.*?)?)?$/)) {
        (label && match[1]) && this.warning(`label ${label} at offset ${this.offset} has been redefine, the previous has been ignored`);
        l && (label = l);
        code = match[2];
        para = match[3];
      }
      else if (match = code_string.match(/^\s*(%\w+)(\s+.*?)?/)) {
        code = match[1];
        para = match[2];
      }
      else {
        this.fatal("parse error")
      }

      if (label) {
        if (!this.code[this.segment]['label'][label]) {
          this.code[this.segment]['label'][label] = this.offset
        }
        else {
          if (this.code[this.segment]['label'][label] != this.offset)
            {this.error(`label ${label} redefine`)}
        }
      }

      if (code) {
        this.code[this.segment]['code'].push(new SingletonCode(code,para,label,comment))
        code.apply(this)
        this.offset += code.bytes
        code.getready()
        label = void(0)
      }
    });

    if (this.errorflag) {return}

    // expression and binary

    for (let key in this.code) {
      if (this.code.hasOwnProperty(key)) {
        let value = this.code[key];
        value['code'].forEach(code => {
          let message = code.getready()
          if (message) {
            thie.error(message)
          }
        })
      }
    }

    if (this.errorflag) {return}

    return this.code

  }

  // exception deal
  warning(info) {this.errorlist.push(`warning[${this.line + 1}]: ${info}`)}
  error(info) {this.errorlist.push(`error[${this.line + 1}]: ${info}`);this.errorflag = true}
  fatal(info) {this.errorlist.push(`fatal[${this.line + 1}]: ${info}`);throw `fatal error [${this.line + 1}]: ${info} ;then Assembler halt`}
}

module.exports = as
